### PR TITLE
Remove the unnecessary "() =>" in `def iteratorSeq` 

### PR DESCRIPTION
**Pull request by AnishShah**
I made an abstract class as @szymonm said.

```
abstract class NodeStream extends Iterable[NodeIdEdgesMaxId] with Closeable {..}
```
Then, I used it in AdjacencyListGraphReader.scala & ListOfEdgesGraphReader.scala.

I implemened the `close()` method to close the file.

Is this correct? I haven't removed "() => " in `def iteratorSeq`.  I will do it after reviews on this.

### CODE REVIEW

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/NodeStream.scala [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright 2014 Twitter, Inc.
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/NodeStream.scala]: Pls use 2015 on all new files

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/NodeStream.scala [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package com.twitter.cassovary.util.io
+
+import com.twitter.cassovary.graph.NodeIdEdgesMaxId
+import java.io.Closeable
+
+abstract class NodeStream extends Iterable[NodeIdEdgesMaxId] with Closeable{
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/NodeStream.scala]: why not a trait?

trait NodeStream extends Iterable[NodeIdEdgesMaxId] with Closeable

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/AdjacencyListGraphReader.scala [hunk 3]
```diff
@@ -81,18 +81,20 @@ class AdjacencyListGraphReader[T] (
       .map{x => {lastLineParsed += 1; x}}
     private val holder = NodeIdEdgesMaxId(-1, null, -1)
 
-    override def hasNext: Boolean = lines.hasNext
+    override def iterator: Iterator[NodeIdEdgesMaxId] = new Iterator[NodeIdEdgesMaxId] {
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/AdjacencyListGraphReader.scala]: no need to specify type of iterator here as it is so clear.

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/ListOfEdgesGraphReader.scala [hunk 4]
```diff
@@ -114,25 +114,32 @@ class ListOfEdgesGraphReader[T](
...
+        }
       }
     }
+
+    override def close(): Unit = {
+      Source.fromFile(filename).close()
+    }
   }
 
   def oneShardReader(filename: String): Iterator[NodeIdEdgesMaxId] = {
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/ListOfEdgesGraphReader.scala]: correct me @szymonm if I am wrong, but we want to change the return value here to NodeStream. Then call NodeStream.close() after all the lines have been read during the array construction. 

And yes @AnishShah pls remove the () => everywhere after this change.

**Comment by AnishShah** [comment]
@pankajgupta Thanks, I will fix this in sometime.

**Comment by AnishShah** [comment]
@pankajgupta @szymonm -  Please check the last commit. I have removed `() =>` from GraphReader. 
On Line 65 & 74
```
iteratorSeq.map(iter => {() => iter.iterator}).toSeq
```
Here, iter.iterator is returning empty. 
I haven't removed `() =>` from all the files. But this should still work, right? I don't know why it is returning empty.

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReader.scala [hunk 5]
```diff
@@ -62,7 +62,7 @@ trait GraphReader[T] {
    * Create an `ArrayBasedDirectedGraph`
    */
   def toArrayBasedDirectedGraph() = {
-    ArrayBasedDirectedGraph(iteratorSeq, parallelismLimit, storedGraphDir)
+    ArrayBasedDirectedGraph(iteratorSeq.map(iter => {() => iter.iterator}).toSeq, parallelismLimit, storedGraphDir)
```
Comment (AnishShah) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReader.scala]: Here, `iter.iterator` is returning `empty`. I don't know why it is returning empty. It should return `Iterator[NodeIdEdgesMaxIs]`

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReader.scala [hunk 6]
```diff
@@ -39,7 +39,7 @@ trait GraphReader[T] {
   /**
    * Should return a sequence of iterators over `NodeIdEdgesMaxId` objects
    */
-  def iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]]
+  def iteratorSeq: Iterable[NodeStream]
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReader.scala]: should be Seq[NodeStream]

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReaderFromDirectory.scala [hunk 7]
```diff
@@ -35,12 +34,12 @@ trait GraphReaderFromDirectory[T] extends GraphReader[T] {
...
    * Returns a reader for a given file (shard).
    */
-  def oneShardReader(filename : String) : Iterator[NodeIdEdgesMaxId]
+  def oneShardReader(filename : String) : NodeStream
 
   /**
    * Should return a sequence of iterators over NodeIdEdgesMaxId objects
    */
-  def iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]] = {
+  def iteratorSeq: Iterable[NodeStream] = {
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReaderFromDirectory.scala]: Seq[NodeStream] is fine

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReader.scala (hunk 5 above)
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReader.scala]: what do you mean 'empty'? in which test does it return empty?

**Comment by AnishShah** [comment]
@pankajgupta @szymonm 

I think there's a bug in `SharedArrayBasedDirectedGraph.scala` because of which this is failing.

The problem is, we are reading the same iterator twice (first in readNumOfEdgesAndMaxId, and second in `readOutEdges`) .

For eg:
```
scala> val list = List(1, 2)
list: List[Int] = List(1, 2)

scala> val it = list.iterator
it: Iterator[Int] = non-empty iterator

scala> val it1 = it
it1: Iterator[Int] = non-empty iterator

scala> val it2 = it
it2: Iterator[Int] = non-empty iterator

scala> it1.next()
res2: Int = 1

scala> it1.next()
res3: Int = 2

scala> it1.next()
java.util.NoSuchElementException: next on empty iterator
  at scala.collection.Iterator$$anon$2.next(Iterator.scala:39)
  at scala.collection.Iterator$$anon$2.next(Iterator.scala:37)
  at scala.collection.LinearSeqLike$$anon$1.next(LinearSeqLike.scala:47)
  ... 43 elided

scala> it2.next()
java.util.NoSuchElementException: next on empty iterator
  at scala.collection.Iterator$$anon$2.next(Iterator.scala:39)
  at scala.collection.Iterator$$anon$2.next(Iterator.scala:37)
  at scala.collection.LinearSeqLike$$anon$1.next(LinearSeqLike.scala:47)
  ... 43 elided
``` 

I haven't made any changes in `SharedArrayBasedDirectedGraph.scala` and when I added `log.info()`  after `readNumOfEdgesAndMaxId` and it said `iteratorFunc as empty iterator`. 
Please correct me if I'm wrong. (I don't know how it was working properly till now.)

**Comment by pankajgupta** [comment]
Oh I see -- yes SharedArrayBasedDirectedGraph is reading it twice. Earlier the () => Iterator type gave a new iterator both times.
So what we need to do is to change Seq[() => Iterator] to Seq[NodeStream] and replace the iteratorFunc() to iteratorFunc.iterator

**Comment by AnishShah** [comment]
There's another problem then.. If we change `Seq[() => Iterator]` to `Seq[NodeStream]`, then in TestGraph.scala, we are creating graphs using `NodeIdEdgesMaxId`. But, we can't create using `NodeStream` because it is just a trait.

**Comment by pankajgupta** [comment]
Aah, we missed that earlier. Actually, seems to me that it will be better to delete NodeStream, go back to NodeIdEdgesMaxId and change Seq[() => Iterator] to Seq[Iterable[NodeIdEdgesMaxId]].

The close() should be handled by the files providing implementation of OneShardReader that actually use and read from Source -- i.e., in ListOfEdgesGraphReader and AdjacencyListGraphReader by closing the file they have opened. 

@szymonm since you had suggested the NodeStream extend Closeable, is this ok by you? doesn't seem like NodeStream needs to be provided only by a file source.

**Comment by AnishShah** [comment]
I'm not able to understand why it is still giving empty iterator even though we are using `Iterable[NodeIdEdgesMaxId]` in `SharedArrayBasedDirectedGraph.scala`. Any help? @pankajgupta @szymonm

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/AdjacencyListGraphReader.scala [hunk 8]
```diff
@@ -73,7 +73,7 @@ class AdjacencyListGraphReader[T] (
    * @param filename Name of file to read from
    */
   private class OneShardReader(filename: String, nodeNumberer: NodeNumberer[T])
-    extends Iterator[NodeIdEdgesMaxId] {
+    extends Iterable[NodeIdEdgesMaxId] {
 
     private val outEdgePattern = ("""^(\w+)""" + separator + """(\d+)""").r
     var lastLineParsed = 0
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/AdjacencyListGraphReader.scala]: move lastLineParsed, lines, holder to inside the new Iterator block. That's why you are seeing empty iterator exception.

File: cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDirectedGraph.scala [hunk 9]
```diff
@@ -144,7 +144,7 @@ object ArrayBasedDirectedGraph {
         var id = 0
         var edgesLength = 0
 
-        val iterator = iteratorFunc()
+        val iterator = iteratorFunc
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDirectedGraph.scala]: just change iteratorFunc variable to iterator in the method parameters

File: cassovary-core/src/main/scala/com/twitter/cassovary/graph/SharedArrayBasedDirectedGraph.scala [hunk 10]
```diff
@@ -66,10 +66,10 @@ object SharedArrayBasedDirectedGraph {
      */
     log.info("read out num of edges and max id from files in parallel")
     val futures = statsReceiver.time("graph_dump_read_out_num_of_edge_and_max_id_parallel") {
-      def readNumOfEdgesAndMaxId(iteratorFunc: () => Iterator[NodeIdEdgesMaxId]) =
+      def readNumOfEdgesAndMaxId(iteratorFunc: Iterable[NodeIdEdgesMaxId]) =
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/graph/SharedArrayBasedDirectedGraph.scala]: change iteratorFunc to some other name: edgesIterable or something else you like

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/AdjacencyListGraphReader.scala [hunk 11]
```diff
@@ -81,18 +81,20 @@ class AdjacencyListGraphReader[T] (
       .map{x => {lastLineParsed += 1; x}}
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/AdjacencyListGraphReader.scala]: close the file at the end too

**Comment by AnishShah** [comment]
Thanks @pankajgupta!

File: cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDirectedGraph.scala [hunk 12]
```diff
@@ -50,7 +50,7 @@ private case class NodesMaxIds(nodesInOnePart: Seq[Node],
                                maxIdInPart: Int, nodeWithOutEdgesMaxIdInPart: Int)
 
 object ArrayBasedDirectedGraph {
-  def apply(iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]], parallelismLimit: Int,
+  def apply(iteratorSeq: Seq[Iterable[NodeIdEdgesMaxId]], parallelismLimit: Int,
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDirectedGraph.scala]: change variable name to iterableSeq

File: cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDirectedGraph.scala [hunk 13]
```diff
@@ -58,14 +58,14 @@ object ArrayBasedDirectedGraph {
   }
 
   @VisibleForTesting
-  def apply(iteratorFunc: () => Iterator[NodeIdEdgesMaxId],
+  def apply(iteratorFunc: Iterable[NodeIdEdgesMaxId],
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDirectedGraph.scala]: change variable name to iterable

File: cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDirectedGraph.scala [hunk 14]
```diff
@@ -58,14 +58,14 @@ object ArrayBasedDirectedGraph {
...
             storedGraphDir: StoredGraphDir): ArrayBasedDirectedGraph = apply(Seq(iteratorFunc), 1, storedGraphDir)
 
   /**
    * Constructs array based directed graph
    */
   private class ArrayBasedDirectedGraphConstructor(
-    iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]],
+    iteratorSeq: Seq[Iterable[NodeIdEdgesMaxId]],
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDirectedGraph.scala]: change variable name to iterableSeq

File: cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDirectedGraph.scala [hunk 15]
```diff
@@ -109,7 +109,7 @@ object ArrayBasedDirectedGraph {
      * In every node only edges directly read from input are set.
      * @return Future with read edges of type `Buffer[Seq[Node]]`, max node id and nodeWithOutEdgesMaxId
      */
-    private def fillOutEdges(iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]], storedGraphDir: StoredGraphDir):
+    private def fillOutEdges(iteratorSeq: Seq[Iterable[NodeIdEdgesMaxId]], storedGraphDir: StoredGraphDir):
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDirectedGraph.scala]: change variable name to iterableSeq

File: cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraph.scala [hunk 16]
```diff
@@ -29,7 +29,7 @@ class ArrayBasedDynamicDirectedGraph(val storedGraphDir: StoredGraphDir)
 
   private var _nodeCount = 0
 
-  def this(dataIterator: Iterator[NodeIdEdgesMaxId],
+  def this(dataIterator: Iterable[NodeIdEdgesMaxId],
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraph.scala]: change variable name to dataIterable

File: cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraph.scala [hunk 17]
```diff
@@ -44,9 +44,9 @@ class ArrayBasedDynamicDirectedGraph(val storedGraphDir: StoredGraphDir)
     }
   }
 
-  def this(iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]],
+  def this(iteratorSeq: Seq[Iterable[NodeIdEdgesMaxId]],
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraph.scala]: change variable name to iterableSeq

File: cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraph.scala [hunk 18]
```diff
@@ -44,9 +44,9 @@ class ArrayBasedDynamicDirectedGraph(val storedGraphDir: StoredGraphDir)
...
            storedGraphDir: StoredGraphDir) {
-    this((iteratorSeq.view flatMap { _()}).iterator, storedGraphDir)
+    this(iteratorSeq flatMap { iterator => iterator } , storedGraphDir)
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraph.scala]: iterableSeq.flatten should work

File: cassovary-core/src/main/scala/com/twitter/cassovary/graph/SharedArrayBasedDirectedGraph.scala [hunk 19]
```diff
@@ -46,7 +46,7 @@ object SharedArrayBasedDirectedGraph {
    * @param storedGraphDir the direction of the graph to be built
    * @param numOfShards specifies the number of shards to use in creating shared array
    */
-  def apply(iteratorSeq: Seq[ () => Iterator[NodeIdEdgesMaxId] ], executorService: ExecutorService,
+  def apply(iteratorSeq: Seq[Iterable[NodeIdEdgesMaxId] ], executorService: ExecutorService,
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/graph/SharedArrayBasedDirectedGraph.scala]: change variable name to iterableSeq

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/AdjacencyListGraphReader.scala [hunk 20]
```diff
@@ -73,26 +73,29 @@ class AdjacencyListGraphReader[T] (
...
 
-    private val outEdgePattern = ("""^(\w+)""" + separator + """(\d+)""").r
-    var lastLineParsed = 0
-    private val lines = Source.fromFile(filename).getLines()
-      .map{x => {lastLineParsed += 1; x}}
-    private val holder = NodeIdEdgesMaxId(-1, null, -1)
+    override def iterator = new Iterator[NodeIdEdgesMaxId] {
 
-    override def hasNext: Boolean = lines.hasNext
+      private val outEdgePattern = ("""^(\w+)""" + separator + """(\d+)""").r
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/AdjacencyListGraphReader.scala]: you can keep immutables (such as definition of patterns) outside the iterator if you want

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/AdjacencyListGraphReader.scala [hunk 21]
```diff
@@ -73,26 +73,29 @@ class AdjacencyListGraphReader[T] (
...
+      var lastLineParsed = 0
+      private val lines = Source.fromFile(filename).getLines()
+        .map{x => {lastLineParsed += 1; x}}
+      Source.fromFile(filename).close()
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/AdjacencyListGraphReader.scala]: this is dangerous to do here as lines is an iterator (that has not been traversed yet). close the file in hasNext
private val src = Source.fromFile(filename)
private val lines = src.getLines().map ....
def hasNext = {
val h = lines.hasNext
if (!h) src.close() //closing an already closed stream does nothing
h
}

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReader.scala [hunk 22]
```diff
@@ -39,7 +39,7 @@ trait GraphReader[T] {
   /**
    * Should return a sequence of iterators over `NodeIdEdgesMaxId` objects
    */
-  def iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]]
+  def iteratorSeq: Seq[Iterable[NodeIdEdgesMaxId]]
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReader.scala]: change var name to iterableSeq

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReaderFromDirectory.scala [hunk 23]
```diff
@@ -13,8 +13,8 @@
  */
 package com.twitter.cassovary.util.io
 
-import com.twitter.cassovary.graph.NodeIdEdgesMaxId
 import java.io.File
+import com.twitter.cassovary.graph.NodeIdEdgesMaxId
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReaderFromDirectory.scala]: maintain alpha imports

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReaderFromDirectory.scala [hunk 24]
```diff
@@ -35,12 +35,12 @@ trait GraphReaderFromDirectory[T] extends GraphReader[T] {
...
    * Returns a reader for a given file (shard).
    */
-  def oneShardReader(filename : String) : Iterator[NodeIdEdgesMaxId]
+  def oneShardReader(filename : String) : Iterable[NodeIdEdgesMaxId]
 
   /**
    * Should return a sequence of iterators over NodeIdEdgesMaxId objects
    */
-  def iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]] = {
+  def iteratorSeq: Seq[Iterable[NodeIdEdgesMaxId]] = {
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReaderFromDirectory.scala]: change var name to iterableSeq

File: cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReaderFromDirectory.scala [hunk 25]
```diff
@@ -55,8 +55,8 @@ trait GraphReaderFromDirectory[T] extends GraphReader[T] {
         None
       }
     })
-    validFiles.map({ filename =>
-    {() => oneShardReader(directory + "/" + filename)}
-    }).toSeq
+    validFiles.map(filename =>
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/util/io/GraphReaderFromDirectory.scala]: prefer () to {} in multi-line block

File: cassovary-core/src/test/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraphSpec.scala [hunk 26]
```diff
@@ -8,9 +8,9 @@ class ArrayBasedDynamicDirectedGraphSpec extends WordSpec with ShouldMatchers wi
   val graphDirections = List(StoredGraphDir.OnlyIn, StoredGraphDir.OnlyOut, StoredGraphDir.BothInOut,
                              StoredGraphDir.Mutual) // Bipartitie is not supported
 
-  def builder(iteratorFunc: () => Iterator[NodeIdEdgesMaxId],
+  def builder(iteratorFunc: Iterable[NodeIdEdgesMaxId],
```
Comment (pankajgupta) [inline cassovary-core/src/test/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraphSpec.scala]: change var name to iterableSeq

**Comment by AnishShah** [comment]
I guess, Something is wrong with Travis.

```
$ sbt ++$TRAVIS_SCALA_VERSION test
Detected sbt version 0.13.7
Downloading sbt launcher for 0.13.7:
  From  http://typesafe.artifactoryonline.com/typesafe/ivy-releases/org.scala-sbt/sbt-launch/0.13.7/sbt-launch.jar
    To  /home/travis/.sbt/launchers/0.13.7/sbt-launch.jar
Download failed. Obtain the jar manually and place it at /home/travis/.sbt/launchers/0.13.7/sbt-launch.jar

The command "sbt ++$TRAVIS_SCALA_VERSION test" exited with 1.

Done. Your build exited with 1.
```

**Comment by AnishShah** [comment]
Sorry for closing & reopening PR twice. Something was wrong with Travis.

All benchmarks and examples works except RandomWalkJava. Though the error was not because of my changes, I have corrected it.

This should be ready for merge, I guess. Tell me if any changes are required.

**Pull request merged**